scipy/stats/_bootstrap.py

```diff
@@ -0,0 +1,393 @@
+        data_iv.append(sample)
+
+    # should we try: statistic(data, axis=axis) here?
+
+    confidence_level_float = float(confidence_level)
+
+    n_resamples_int = int(n_resamples)
+    if n_resamples != n_resamples_int or n_resamples_int <= 0:
+        raise ValueError("`n_resamples` must be a positive integer.")
+
+    methods = {'percentile', 'basic', 'bca'}
+    method = method.lower()
+    if method not in methods:
+        raise ValueError(f"`method` must be in {methods}")
+
+    message = "`method = 'BCa' is only available for one-sample statistics"
+    if n_samples > 1 and method == 'bca':
+        raise ValueError(message)
+
+    random_state = check_random_state(random_state)
+
+    return (data_iv, statistic, axis_int, confidence_level_float,
+            n_resamples_int, method, random_state)
+
+
+fields = ['confidence_interval', 'standard_error']
+BootstrapResult = make_dataclass("BootstrapResult", fields)
+
+
+def bootstrap(data, statistic, *, axis=0, confidence_level=0.95,
+              n_resamples=9999, method='basic', random_state=None):
+    r"""
+    Compute a two-sided bootstrap confidence interval of a statistic.
+
+    When `method` is ``'percentile'``, a bootstrap confidence interval is
+    computed according to the following procedure.
+
+    1. Resample the data: for each sample in `data` and for each of
+       `n_resamples`, take a random sample of the original sample
+       (with replacement) of the same size as the original sample.
+
+    2. Compute the bootstrap distribution of the statistic: for each set of
+       resamples, compute the test statistic.
+
+    3. Determine the confidence interval: find the interval of the bootstrap
+       distribution that is
+
+       - symmetric about the median and
+       - contains `confidence_level` of the resampled statistic values.
+
+    While the ``'percentile'`` method is the most intuitive, it is rarely
+    used in practice. Two more common methods are available, ``'basic'``
+    ('reverse percentile') and ``'BCa'`` ('bias-corrected and accelerated');
+    they differ in how step 3 is performed.
+
+    If the samples in `data` are  taken at random from their respective
+    distributions :math:`n` times, the confidence interval returned by
+    `bootstrap` will contain the true value of the statistic for those
+    distributions approximately `confidence_level`:math:`\, \times \, n` times.
+
+    Parameters
+    ----------
+    data : sequence of array-like
+         Each element of data is a sample from an underlying distribution.
+    statistic : callable
+        Statistic for which the confidence interval is to be calculated.
+        `statistic` must be a callable that accepts ``len(data)`` samples
+        as separate arguments and returns the resulting statistic.
+        `statistic` must also accept a keyword argument `axis` and be
+        vectorized to compute the statistic along the provided `axis`.
+    axis : int, optional
```

Comment:
The mandatory axis seems a bit limiting. What if I wanted to bootstrap regression estimates using `np.linalg.lstsq`?  Could this be optional so that if `axis=None` then it isn't used in the call?

Reply:
The intent is that you would wrap your function. I will check but i believe it's always called with axis=-1. We could specify that the statistic must be calculated along the last axis, instead.

I suppose if axis=None we could just not use it, sure.

Reply:
Maybe an example in the docstring of this case would be useful in this case?

Reply:
I think I already added one:
![image](https://user-images.githubusercontent.com/6570539/118151354-64266100-b3c8-11eb-9054-ad6c804197d9.png)

Reply:
I was thinking of an example where `np.apply_along_axis` would not obviously work (I think `lstsq` falls into this category).

Reply:
We'll have to decide whether that is common enough use-case for the main documentation or whether it should be added in a tutorial - which we would like to add in a follow-up. If you post a gist of the basic regression estimate code, I can take a look.